File: push/src/main/java/com/xiaomi/channel/commonutils/android/AppInfoUtils.java
package com.xiaomi.channel.commonutils.android;

import android.app.ActivityManager;
import android.app.AppOpsManager;
import android.app.NotificationManager;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Process;
import android.text.TextUtils;
import android.util.Base64;
import com.google.protobuf.micro.CodedOutputStreamMicro;
import com.xiaomi.channel.commonutils.logger.MyLog;
import com.xiaomi.channel.commonutils.reflect.JavaCalls;
import com.xiaomi.push.service.MIPushAccount;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/channel/commonutils/android/AppInfoUtils.class */
public class AppInfoUtils {
    private static final String ANDROID_PERMISSION_PREF = "android.permission.";
    public static final int PATTERN = 100000;
    public static final String SEPARATE_ITEM = "#";
    private static final String TAG = "AppInfoUtils.";

    /* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/channel/commonutils/android/AppInfoUtils$AppNotificationOp.class */
    public enum AppNotificationOp {
        UNKNOWN(0),
        ALLOWED(1),
        NOT_ALLOWED(2);

        public static final int MASK = 3;
        private final int value;

        AppNotificationOp(int i) {
            this.value = i;
        }

        public static AppNotificationOp findByValue(int i) {
            switch (i) {
                case 0:
                    return UNKNOWN;
                case 1:
                    return ALLOWED;
                case 2:
                    return NOT_ALLOWED;
                default:
                    return null;
            }
        }

        public int getValue() {
            return this.value;
        }
    }

    private AppInfoUtils() {
    }

    private static Integer coerceInteger(Object value) {
        if (value instanceof Integer) {
            return (Integer) value;
        }
        if (value instanceof Number) {
            return Integer.valueOf(((Number) value).intValue());
        }
        return null;
    }

    private static Object invokeCheckOpNoThrow(Object appOpsService, int op, int uid, String packageName) {
        if (appOpsService == null) {
            return null;
        }
        try {
            Method method = appOpsService.getClass().getMethod("checkOpNoThrow", Integer.TYPE, Integer.TYPE, String.class);
            method.setAccessible(true);
            return method.invoke(appOpsService, Integer.valueOf(op), Integer.valueOf(uid), packageName);
        } catch (Throwable th) {
            MyLog.w("checkOpNoThrow invoke error service=" + appOpsService.getClass().getName() + " " + th);
            return null;
        }
    }

    private static AppNotificationOp areNotificationsEnabled(Context context, ApplicationInfo applicationInfo) {
        Boolean boolValueOf;
        int i = Build.VERSION.SDK_INT;
        if (applicationInfo == null || i < 24) {
            return AppNotificationOp.UNKNOWN;
        }
        try {
            if (applicationInfo.packageName.equals(context.getPackageName())) {
                boolValueOf = Boolean.valueOf(((NotificationManager) context.getSystemService("notification")).areNotificationsEnabled());
            } else {
                Object objCallMethod = i >= 29 ? JavaCalls.callMethod(context.getSystemService("notification"), "getService", new Object[0]) : context.getSystemService(MIPushAccount.PREF_KEY_SECURITY);
                boolValueOf = null;
                if (objCallMethod != null) {
                    boolValueOf = (Boolean) JavaCalls.callMethodOrThrow(objCallMethod, "areNotificationsEnabledForPackage", applicationInfo.packageName, Integer.valueOf(applicationInfo.uid));
                }
            }
            if (boolValueOf != null) {
                return boolValueOf.booleanValue() ? AppNotificationOp.ALLOWED : AppNotificationOp.NOT_ALLOWED;
            }
        } catch (Exception e) {
            MyLog.w("are notifications enabled error " + e);
        }
        return AppNotificationOp.UNKNOWN;
    }

    public static boolean checkSelfPermission(Context context, String str) {
        return context.getPackageManager().checkPermission(str, context.getPackageName()) == 0;
    }

    public static String convertPermissionString(String[] strArr) {
        int i;
        boolean z;
        AppPermissionType[] appPermissionTypeArrValues = AppPermissionType.values();
        byte[] bArr = new byte[(int) Math.ceil(((double) appPermissionTypeArrValues.length) / 8.0d)];
        int i2 = -1;
        if (strArr == null) {
            MyLog.v("AppInfoUtils.: no permissions");
            return "";
        }
        int length = strArr.length;
        int i3 = 0;
        while (i3 < length) {
            String str = strArr[i3];
            int i4 = i2;
            if (!TextUtils.isEmpty(str)) {
                if (str.startsWith(ANDROID_PERMISSION_PREF)) {
                    int i5 = 0;
                    while (true) {
                        i = i2;
                        z = false;
                        if (i5 >= appPermissionTypeArrValues.length) {
                            break;
                        }
                        if (TextUtils.equals(ANDROID_PERMISSION_PREF + appPermissionTypeArrValues[i5].name(), str)) {
                            z = true;
                            i = i5;
                            break;
                        }
                        i5++;
                    }
                    i4 = i;
                    if (z) {
                        i4 = i;
                        if (i != -1) {
                            bArr[i / 8] = (byte) (bArr[i / 8] | (1 << (7 - (i % 8))));
                            i4 = i;
                        }
                    }
                } else {
                    i4 = i2;
                }
            }
            i3++;
            i2 = i4;
        }
        return new String(Base64.encode(bArr, 0));
    }

    public static Drawable getAppIconDrawable(Context context, String str) {
        ApplicationInfo applicationInfo = getApplicationInfo(context, str);
        Drawable drawableLoadLogo = null;
        Drawable drawable = null;
        if (applicationInfo != null) {
            try {
                Drawable drawableLoadIcon = applicationInfo.loadIcon(context.getPackageManager());
                drawableLoadLogo = drawableLoadIcon;
                if (drawableLoadIcon == null) {
                    drawableLoadLogo = drawableLoadIcon;
                    if (Build.VERSION.SDK_INT >= 9) {
                        drawable = drawableLoadIcon;
                        drawableLoadLogo = applicationInfo.loadLogo(context.getPackageManager());
                    }
                }
            } catch (Exception e) {
                MyLog.w("get app icon drawable failed, " + e);
                drawableLoadLogo = drawable;
            }
        }
        if (drawableLoadLogo == null) {
            drawableLoadLogo = new ColorDrawable(0);
        }
        return drawableLoadLogo;
    }

    public static int getAppIconId(Context context, String str) {
        ApplicationInfo applicationInfo = getApplicationInfo(context, str);
        int i = 0;
        if (applicationInfo != null) {
            int i2 = applicationInfo.icon;
            i = i2;
            if (i2 == 0) {
                i = i2;
                if (Build.VERSION.SDK_INT >= 9) {
                    i = applicationInfo.logo;
                }
            }
        }
        return i;
    }

    public static String getAppLabel(Context context, String str) {
        String string;
        try {
            PackageManager packageManager = context.getPackageManager();
            PackageInfo packageInfo = packageManager.getPackageInfo(str, 0);
            string = str;
            if (packageInfo != null) {
                ApplicationInfo applicationInfo = packageInfo.applicationInfo;
                string = str;
                if (applicationInfo != null) {
                    string = packageManager.getApplicationLabel(applicationInfo).toString();
                }
            }
        } catch (PackageManager.NameNotFoundException e) {
            MyLog.e(e);
            string = str;
        }
        return string;
    }

    public static AppNotificationOp getAppNotificationOp(Context context, String str, boolean z) {
        ApplicationInfo applicationInfo = null;
        AppNotificationOp appNotificationOpAreNotificationsEnabled = AppNotificationOp.UNKNOWN;
        if (context == null || TextUtils.isEmpty(str) || Build.VERSION.SDK_INT < 19) {
            return AppNotificationOp.UNKNOWN;
        }
        try {
            applicationInfo = str.equals(context.getPackageName()) ? context.getApplicationInfo() : context.getPackageManager().getApplicationInfo(str, 0);
            appNotificationOpAreNotificationsEnabled = areNotificationsEnabled(context, applicationInfo);
        } catch (Throwable th) {
            MyLog.w("get app op error " + th);
        }
        if (appNotificationOpAreNotificationsEnabled != AppNotificationOp.UNKNOWN) {
            return appNotificationOpAreNotificationsEnabled;
        }
        if (applicationInfo == null) {
            return AppNotificationOp.UNKNOWN;
        }
        Integer num = coerceInteger(JavaCalls.getStaticField((Class<? extends Object>) AppOpsManager.class, "OP_POST_NOTIFICATION"));
        if (num == null) {
            return AppNotificationOp.UNKNOWN;
        }
        Integer num2 = coerceInteger(
            invokeCheckOpNoThrow(
                context.getSystemService("appops"),
                num.intValue(),
                applicationInfo.uid,
                str
            )
        );
        Integer num3 = coerceInteger(JavaCalls.getStaticField((Class<? extends Object>) AppOpsManager.class, "MODE_ALLOWED"));
        Integer num4 = coerceInteger(JavaCalls.getStaticField((Class<? extends Object>) AppOpsManager.class, "MODE_IGNORED"));
        MyLog.i(String.format("get app mode %s|%s|%s", num2, num3, num4));
        Integer num5 = num3;
        if (num3 == null) {
            num5 = 0;
        }
        Integer num6 = num4;
        if (num4 == null) {
            num6 = 1;
        }
        if (num2 != null) {
            if (z) {
                return !num2.equals(num6) ? AppNotificationOp.ALLOWED : AppNotificationOp.NOT_ALLOWED;
            }
            return num2.equals(num5) ? AppNotificationOp.ALLOWED : AppNotificationOp.NOT_ALLOWED;
        }
        return AppNotificationOp.UNKNOWN;
    }

    public static String getAppPermissionBase64Str(Context context, String str) {
        try {
            return convertPermissionString(context.getPackageManager().getPackageInfo(str, CodedOutputStreamMicro.DEFAULT_BUFFER_SIZE).requestedPermissions);
        } catch (PackageManager.NameNotFoundException e) {
            MyLog.e(e.toString());
            return "";
        }
    }

    private static ApplicationInfo getApplicationInfo(Context context, String str) {
        ApplicationInfo applicationInfo;
        if (str.equals(context.getPackageName())) {
            applicationInfo = context.getApplicationInfo();
        } else {
            try {
                applicationInfo = context.getPackageManager().getApplicationInfo(str, 0);
            } catch (PackageManager.NameNotFoundException e) {
                MyLog.w("not found app info " + str);
                applicationInfo = null;
            }
        }
        return applicationInfo;
    }

    public static int getArchiveVersionCode(Context context, String str) {
        try {
            PackageInfo packageArchiveInfo = context.getPackageManager().getPackageArchiveInfo(str, 1);
            if (packageArchiveInfo == null) {
                return 0;
            }
            return (int) packageArchiveInfo.getLongVersionCode();
        } catch (Exception e) {
            return 0;
        }
    }

    public static String getForegroundApp(Context context) {
        return null;
    }

    public static String getProcessName(Context context) {
        List<ActivityManager.RunningAppProcessInfo> runningAppProcesses;
        if (context == null || (runningAppProcesses = ((ActivityManager) context.getSystemService("activity")).getRunningAppProcesses()) == null) {
            return null;
        }
        int iMyPid = Process.myPid();
        for (ActivityManager.RunningAppProcessInfo runningAppProcessInfo : runningAppProcesses) {
            if (runningAppProcessInfo.pid == iMyPid) {
                return runningAppProcessInfo.processName;
            }
        }
        return null;
    }

    public static String getRunningAppPkgNames(Context context) {
        List<ActivityManager.RunningAppProcessInfo> runningAppProcesses = ((ActivityManager) context.getSystemService("activity")).getRunningAppProcesses();
        ArrayList<String> arrayList = new ArrayList<>();
        StringBuilder sb = new StringBuilder();
        if (runningAppProcesses != null && runningAppProcesses.size() > 0) {
            Iterator<ActivityManager.RunningAppProcessInfo> it = runningAppProcesses.iterator();
            while (it.hasNext()) {
                String[] strArr = it.next().pkgList;
                for (int i = 0; strArr != null && i < strArr.length; i++) {
                    if (!arrayList.contains(strArr[i])) {
                        arrayList.add(strArr[i]);
                        if (arrayList.size() == 1) {
                            sb.append(((String) arrayList.get(0)).hashCode() % PATTERN);
                        } else {
                            sb.append("#");
                            sb.append(strArr[i].hashCode() % PATTERN);
                        }
                    }
                }
            }
        }
        return sb.toString();
    }

    public static Signature[] getSignature(Context context, String str) {
        try {
            PackageInfo packageArchiveInfo = context.getPackageManager().getPackageArchiveInfo(str, PackageManager.GET_SIGNING_CERTIFICATES);
            if (packageArchiveInfo == null || packageArchiveInfo.signingInfo == null) {
                return null;
            }
            return packageArchiveInfo.signingInfo.hasMultipleSigners()
                ? packageArchiveInfo.signingInfo.getApkContentsSigners()
                : packageArchiveInfo.signingInfo.getSigningCertificateHistory();
        } catch (Exception e) {
            return null;
        }
    }

    public static int getVersionCode(Context context, String str) {
        PackageInfo packageInfo;
        try {
            packageInfo = context.getPackageManager().getPackageInfo(str, 16384);
        } catch (Exception e) {
            MyLog.e(e);
            packageInfo = null;
        }
        if (packageInfo == null) {
            return 0;
        }
        return (int) packageInfo.getLongVersionCode();
    }

    public static String getVersionName(Context context, String str) {
        PackageInfo packageInfo;
        try {
            packageInfo = context.getPackageManager().getPackageInfo(str, 16384);
        } catch (Exception e) {
            MyLog.e(e);
            packageInfo = null;
        }
        return packageInfo != null ? packageInfo.versionName : "1.0";
    }

    public static boolean isAppMainProc(Context context) {
        List<ActivityManager.RunningAppProcessInfo> runningAppProcesses = ((ActivityManager) context.getSystemService("activity")).getRunningAppProcesses();
        if (runningAppProcesses == null || runningAppProcesses.size() < 1) {
            return false;
        }
        for (ActivityManager.RunningAppProcessInfo runningAppProcessInfo : runningAppProcesses) {
            if (runningAppProcessInfo.pid == Process.myPid() && runningAppProcessInfo.processName.equals(context.getPackageName())) {
                return true;
            }
        }
        return false;
    }

    public static boolean isAppRunning(Context context, String str) {
        List<ActivityManager.RunningAppProcessInfo> runningAppProcesses = ((ActivityManager) context.getSystemService("activity")).getRunningAppProcesses();
        if (runningAppProcesses == null) {
            return false;
        }
        Iterator<ActivityManager.RunningAppProcessInfo> it = runningAppProcesses.iterator();
        while (it.hasNext()) {
            if (Arrays.asList(it.next().pkgList).contains(str)) {
                return true;
            }
        }
        return false;
    }

    public static boolean isForeground(Context context) {
        return TextUtils.equals(context.getPackageName(), getForegroundApp(context));
    }

    public static boolean isPkgInstalled(Context context, String str) {
        PackageInfo packageInfo;
        try {
            packageInfo = context.getPackageManager().getPackageInfo(str, 0);
        } catch (PackageManager.NameNotFoundException e) {
            packageInfo = null;
        }
        return packageInfo != null;
    }
}

File: legacy-runtime/src/main/java/com/xiaomi/channel/commonutils/android/AppInfoUtils.kt
package com.xiaomi.channel.commonutils.android

import android.app.ActivityManager
import android.app.AppOpsManager
import android.app.NotificationManager
import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.pm.Signature
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.os.Build
import android.os.Process
import android.text.TextUtils
import android.util.Base64
import com.xiaomi.channel.commonutils.logger.MyLog
import com.xiaomi.channel.commonutils.reflect.JavaCalls
import java.lang.reflect.Method
import kotlin.math.ceil

/*
 * Current override reference: com.xiaomi.xmsf 0.3.17-20260410000745 (versionCode 1003003000),
 * base.apk sha256 f3d72b6f5e1427ceecd3147a051d58e4dc95bb528397d486658e01cad9f7e590,
 * JADX path: com.xiaomi.xmsf/current/base/sources/com/xiaomi/channel/commonutils/android/AppInfoUtils.java
 */
object AppInfoUtils {
    private const val ANDROID_PERMISSION_PREF = "android.permission."
    private const val LEGACY_SECURITY_SERVICE = "security"
    private const val REQUESTED_PERMISSION_FLAGS = 4096
    const val PATTERN = 100000
    const val SEPARATE_ITEM = "#"
    private const val TAG = "AppInfoUtils."

    enum class AppNotificationOp(val value: Int) {
        UNKNOWN(0),
        ALLOWED(1),
        NOT_ALLOWED(2),
        ;

        companion object {
            const val MASK = 3

            @JvmStatic
            fun findByValue(i: Int): AppNotificationOp? {
                return when (i) {
                    0 -> UNKNOWN
                    1 -> ALLOWED
                    2 -> NOT_ALLOWED
                    else -> null
                }
            }
        }
    }

    private fun coerceInteger(value: Any?): Int? {
        return when (value) {
            is Int -> value
            is Number -> value.toInt()
            else -> null
        }
    }

    private fun invokeCheckOpNoThrow(appOpsService: Any?, op: Int, uid: Int, packageName: String): Any? {
        if (appOpsService == null) {
            return null
        }
        return try {
            val method: Method = appOpsService.javaClass.getMethod(
                "checkOpNoThrow",
                Int::class.javaPrimitiveType,
                Int::class.javaPrimitiveType,
                String::class.java,
            )
            method.isAccessible = true
            method.invoke(appOpsService, op, uid, packageName)
        } catch (throwable: Throwable) {
            MyLog.w("checkOpNoThrow invoke error service=${appOpsService.javaClass.name} $throwable")
            null
        }
    }

    private fun areNotificationsEnabled(
        context: Context,
        applicationInfo: ApplicationInfo?,
    ): AppNotificationOp {
        val sdkInt = Build.VERSION.SDK_INT
        if (applicationInfo == null || sdkInt < 24) {
            return AppNotificationOp.UNKNOWN
        }
        try {
            val enabled = if (applicationInfo.packageName == context.packageName) {
                (context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager).areNotificationsEnabled()
            } else {
                val service = if (sdkInt >= 29) {
                    JavaCalls.callMethod(context.getSystemService(Context.NOTIFICATION_SERVICE), "getService")
                } else {
                    context.getSystemService(LEGACY_SECURITY_SERVICE)
                }
                if (service != null) {
                    JavaCalls.callMethodOrThrow(
                        service,
                        "areNotificationsEnabledForPackage",
                        applicationInfo.packageName,
                        applicationInfo.uid,
                    ) as? Boolean
                } else {
                    null
                }
            }
            if (enabled != null) {
                return if (enabled) AppNotificationOp.ALLOWED else AppNotificationOp.NOT_ALLOWED
            }
        } catch (e: Exception) {
            MyLog.w("are notifications enabled error $e")
        }
        return AppNotificationOp.UNKNOWN
    }

    @JvmStatic
    fun checkSelfPermission(context: Context, str: String): Boolean {
        return context.packageManager.checkPermission(str, context.packageName) == 0
    }

    @JvmStatic
    fun convertPermissionString(strArr: Array<String?>?): String {
        val permissionTypes = AppPermissionType.values()
        val result = ByteArray(ceil(permissionTypes.size / 8.0).toInt())
        var lastIndex = -1
        if (strArr == null) {
            MyLog.v("$TAG: no permissions")
            return ""
        }
        for (permission in strArr) {
            var currentIndex = lastIndex
            if (!TextUtils.isEmpty(permission) && permission!!.startsWith(ANDROID_PERMISSION_PREF)) {
                var found = false
                var matchedIndex = lastIndex
                for (i in permissionTypes.indices) {
                    if (TextUtils.equals(ANDROID_PERMISSION_PREF + permissionTypes[i].name, permission)) {
                        found = true
                        matchedIndex = i
                        break
                    }
                }
                currentIndex = matchedIndex
                if (found && matchedIndex != -1) {
                    val byteIndex = matchedIndex / 8
                    val mask = 1 shl (7 - (matchedIndex % 8))
                    result[byteIndex] = (result[byteIndex].toInt() or mask).toByte()
                }
            }
            lastIndex = currentIndex
        }
        return String(Base64.encode(result, 0))
    }

    @JvmStatic
    fun getAppIconDrawable(context: Context, str: String): Drawable {
        val applicationInfo = getApplicationInfo(context, str)
        var drawable: Drawable? = null
        if (applicationInfo != null) {
            try {
                drawable = applicationInfo.loadIcon(context.packageManager)
                if (drawable == null && Build.VERSION.SDK_INT >= 9) {
                    drawable = applicationInfo.loadLogo(context.packageManager)
                }
            } catch (e: Exception) {
                MyLog.w("get app icon drawable failed, $e")
            }
        }
        return drawable ?: ColorDrawable(0)
    }

    @JvmStatic
    fun getAppIconId(context: Context, str: String): Int {
        val applicationInfo = getApplicationInfo(context, str) ?: return 0
        var icon = applicationInfo.icon
        if (icon == 0 && Build.VERSION.SDK_INT >= 9) {
            icon = applicationInfo.logo
        }
        return icon
    }

    @JvmStatic
    fun getAppLabel(context: Context, str: String): String {
        return try {
            val packageManager = context.packageManager
            val packageInfo = packageManager.getPackageInfo(str, 0)
            val applicationInfo = packageInfo.applicationInfo
            if (applicationInfo != null) {
                packageManager.getApplicationLabel(applicationInfo).toString()
            } else {
                str
            }
        } catch (e: PackageManager.NameNotFoundException) {
            MyLog.e(e)
            str
        }
    }

    @JvmStatic
    fun getAppNotificationOp(context: Context?, str: String?, z: Boolean): AppNotificationOp {
        var applicationInfo: ApplicationInfo? = null
        var notificationOp = AppNotificationOp.UNKNOWN
        if (context == null || TextUtils.isEmpty(str) || Build.VERSION.SDK_INT < 19) {
            return AppNotificationOp.UNKNOWN
        }
        try {
            applicationInfo = if (str == context.packageName) {
                context.applicationInfo
            } else {
                context.packageManager.getApplicationInfo(str!!, 0)
            }
            notificationOp = areNotificationsEnabled(context, applicationInfo)
        } catch (throwable: Throwable) {
            MyLog.w("get app op error $throwable")
        }
        if (notificationOp != AppNotificationOp.UNKNOWN) {
            return notificationOp
        }
        if (applicationInfo == null) {
            return AppNotificationOp.UNKNOWN
        }
        val postNotificationOp =
            coerceInteger(JavaCalls.getStaticField(AppOpsManager::class.java, "OP_POST_NOTIFICATION"))
                ?: return AppNotificationOp.UNKNOWN
        val mode = coerceInteger(
            invokeCheckOpNoThrow(
                context.getSystemService(Context.APP_OPS_SERVICE),
                postNotificationOp,
                applicationInfo.uid,
                str!!,
            )
        )
        val modeAllowed = coerceInteger(
            JavaCalls.getStaticField(AppOpsManager::class.java, "MODE_ALLOWED")
        ) ?: 0
        val modeIgnored = coerceInteger(
            JavaCalls.getStaticField(AppOpsManager::class.java, "MODE_IGNORED")
        ) ?: 1
        MyLog.i(String.format("get app mode %s|%s|%s", mode, modeAllowed, modeIgnored))
        if (mode != null) {
            return if (z) {
                if (mode != modeIgnored) AppNotificationOp.ALLOWED else AppNotificationOp.NOT_ALLOWED
            } else {
                if (mode == modeAllowed) AppNotificationOp.ALLOWED else AppNotificationOp.NOT_ALLOWED
            }
        }
        return AppNotificationOp.UNKNOWN
    }

    @JvmStatic
    fun getAppPermissionBase64Str(context: Context, str: String): String {
        return try {
            convertPermissionString(context.packageManager.getPackageInfo(str, REQUESTED_PERMISSION_FLAGS).requestedPermissions)
        } catch (e: PackageManager.NameNotFoundException) {
            MyLog.e(e.toString())
            ""
        }
    }

    private fun getApplicationInfo(context: Context, str: String): ApplicationInfo? {
        return if (str == context.packageName) {
            context.applicationInfo
        } else {
            try {
                context.packageManager.getApplicationInfo(str, 0)
            } catch (_: PackageManager.NameNotFoundException) {
                MyLog.w("not found app info $str")
                null
            }
        }
    }

    @JvmStatic
    fun getArchiveVersionCode(context: Context, str: String): Int {
        return try {
            val packageArchiveInfo = context.packageManager.getPackageArchiveInfo(str, 1) ?: return 0
            packageArchiveInfo.longVersionCode.toInt()
        } catch (_: Exception) {
            0
        }
    }

    @JvmStatic
    fun getForegroundApp(context: Context): String? {
        return null
    }

    @JvmStatic
    fun getProcessName(context: Context?): String? {
        val runningAppProcesses =
            (context?.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager)?.runningAppProcesses
                ?: return null
        val myPid = Process.myPid()
        for (runningAppProcessInfo in runningAppProcesses) {
            if (runningAppProcessInfo.pid == myPid) {
                return runningAppProcessInfo.processName
            }
        }
        return null
    }

    @JvmStatic
    fun getRunningAppPkgNames(context: Context): String {
        val runningAppProcesses =
            (context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager).runningAppProcesses
        val packageNames = ArrayList<String>()
        val sb = StringBuilder()
        if (runningAppProcesses != null && runningAppProcesses.isNotEmpty()) {
            for (runningAppProcessInfo in runningAppProcesses) {
                val pkgList = runningAppProcessInfo.pkgList
                for (i in pkgList.indices) {
                    if (!packageNames.contains(pkgList[i])) {
                        packageNames.add(pkgList[i])
                        if (packageNames.size == 1) {
                            sb.append(packageNames[0].hashCode() % PATTERN)
                        } else {
                            sb.append(SEPARATE_ITEM)
                            sb.append(pkgList[i].hashCode() % PATTERN)
                        }
                    }
                }
            }
        }
        return sb.toString()
    }

    @JvmStatic
    fun getSignature(context: Context, str: String): Array<Signature>? {
        return try {
            val packageArchiveInfo =
                context.packageManager.getPackageArchiveInfo(str, PackageManager.GET_SIGNING_CERTIFICATES)
                    ?: return null
            val signingInfo = packageArchiveInfo.signingInfo ?: return null
            if (signingInfo.hasMultipleSigners()) {
                signingInfo.apkContentsSigners
            } else {
                signingInfo.signingCertificateHistory
            }
        } catch (_: Exception) {
            null
        }
    }

    @JvmStatic
    fun getVersionCode(context: Context, str: String): Int {
        val packageInfo = try {
            context.packageManager.getPackageInfo(str, 16384)
        } catch (e: Exception) {
            MyLog.e(e)
            null
        }
        return packageInfo?.longVersionCode?.toInt() ?: 0
    }

    @JvmStatic
    fun getVersionName(context: Context, str: String): String {
        val packageInfo: PackageInfo? = try {
            context.packageManager.getPackageInfo(str, 16384)
        } catch (e: Exception) {
            MyLog.e(e)
            null
        }
        return packageInfo?.versionName ?: "1.0"
    }

    @JvmStatic
    fun isAppMainProc(context: Context): Boolean {
        val runningAppProcesses =
            (context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager).runningAppProcesses
                ?: return false
        if (runningAppProcesses.isEmpty()) {
            return false
        }
        for (runningAppProcessInfo in runningAppProcesses) {
            if (runningAppProcessInfo.pid == Process.myPid() && runningAppProcessInfo.processName == context.packageName) {
                return true
            }
        }
        return false
    }

    @JvmStatic
    fun isAppRunning(context: Context, str: String): Boolean {
        val runningAppProcesses =
            (context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager).runningAppProcesses
                ?: return false
        for (runningAppProcessInfo in runningAppProcesses) {
            if (runningAppProcessInfo.pkgList.contains(str)) {
                return true
            }
        }
        return false
    }

    @JvmStatic
    fun isForeground(context: Context): Boolean {
        return TextUtils.equals(context.packageName, getForegroundApp(context))
    }

    @JvmStatic
    fun isPkgInstalled(context: Context, str: String): Boolean {
        val packageInfo = try {
            context.packageManager.getPackageInfo(str, 0)
        } catch (_: PackageManager.NameNotFoundException) {
            null
        }
        return packageInfo != null
    }
}
